action/protocol/staking/handler_stake_migrate.go

```diff
@@ -0,0 +1,181 @@
+	// delete bucket and bucket index
+	if err := csm.delBucketAndIndex(bucket.Owner, bucket.Candidate, bucket.Index); err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to delete bucket for candidate %s", bucket.Candidate.String())
+	}
+
+	// update bucket pool
+	if err := csm.CreditBucketPool(bucket.StakedAmount); err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to update staking bucket pool %s", err.Error())
+	}
+	// update candidate vote
+	weightedVote := p.calculateVoteWeight(bucket, false)
+	if err := cand.SubVote(weightedVote); err != nil {
+		return nil, nil, &handleError{
+			err:           errors.Wrapf(err, "failed to subtract vote for candidate %s", bucket.Candidate.String()),
+			failureStatus: iotextypes.ReceiptStatus_ErrNotEnoughBalance,
+		}
+	}
+	// clear candidate's self stake if the
+	if cand.SelfStakeBucketIdx == bucket.Index {
+		cand.SelfStake = big.NewInt(0)
+		cand.SelfStakeBucketIdx = candidateNoSelfStakeBucketIndex
+	}
+	if err := csm.Upsert(cand); err != nil {
+		return nil, nil, csmErrorToHandleError(cand.GetIdentifier().String(), err)
+	}
+	// update withdrawer balance
+	if err := withdrawer.AddBalance(bucket.StakedAmount); err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to add balance %s", bucket.StakedAmount)
+	}
+	// put updated withdrawer's account state to trie
+	actionCtx := protocol.MustGetActionCtx(ctx)
+	if err := accountutil.StoreAccount(csm.SM(), actionCtx.Caller, withdrawer); err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to store account %s", actionCtx.Caller.String())
+	}
+	// create receipt log
+	actLog := newReceiptLog(p.addr.String(), handleCandidateActivate, protocol.MustGetFeatureCtx(ctx).NewStakingReceiptFormat)
+	actLog.AddTopics(byteutil.Uint64ToBytesBigEndian(bucket.Index), bucket.Candidate.Bytes())
+	actLog.AddAddress(actionCtx.Caller)
+	actLog.SetData(bucket.StakedAmount.Bytes())
+	return actLog, &action.TransactionLog{
+		Type:      iotextypes.TransactionLogType_WITHDRAW_BUCKET,
+		Amount:    bucket.StakedAmount,
+		Sender:    address.StakingBucketPoolAddr,
+		Recipient: actionCtx.Caller.String(),
+	}, nil
+}
+
+func (p *Protocol) createNFTBucket(ctx context.Context, act *action.MigrateStake, amount, duration *big.Int, cand *Candidate, sm protocol.StateManager) (*action.Receipt, error) {
+	ptl, ok := protocol.MustGetRegistry(ctx).Find(executionProtocolID)
+	if !ok {
+		return nil, errors.New("execution protocol is not registered")
+	}
+	exctPtl, ok := ptl.(executionProtocol)
+	contractAddress := p.config.MigrateContractAddress
+	data, err := StakingContractABI.Pack("stake0", duration, cand.GetIdentifier())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to pack data for contract call")
+	}
+	exeAct, err := action.NewExecution(
+		contractAddress,
+		act.Nonce(),
```

Comment:
is nonce important? if not, can we leave it as 0?

Reply:
It's critical to SetNonce of the caller,